In [43]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.impute import SimpleImputer

In [3]:
 data = pd.read_csv("train.csv")
    

In [4]:
data.head()


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [5]:




print(data.columns)
print(data.isna().sum())

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


# now we have to fill the missing values for there are 3  columns having missing values
 * age = 177 - mean() 
    * cabin = 687
       * embarked =  2


In [6]:
#split dataset for rewquired feactures
data.drop(["PassengerId","Name","Ticket","Fare"], axis = 1, inplace = True)

In [7]:
data


Survived  Pclass     Sex   Age  SibSp  Parch Cabin Embarked
0           0       3    male  22.0      1      0   NaN        S
1           1       1  female  38.0      1      0   C85        C
2           1       3  female  26.0      0      0   NaN        S
3           1       1  female  35.0      1      0  C123        S
4           0       3    male  35.0      0      0   NaN        S
..        ...     ...     ...   ...    ...    ...   ...      ...
886         0       2    male  27.0      0      0   NaN        S
887         1       1  female  19.0      0      0   B42        S
888         0       3  female   NaN      1      2   NaN        S
889         1       1    male  26.0      0      0  C148        C
890         0       3    male  32.0      0      0   NaN        Q

[891 rows x 8 columns]

In [8]:
#spliting data into labels and feactures
x = data.drop("Survived", axis = 1)
y= data["Survived"]

In [9]:
x

Pclass     Sex   Age  SibSp  Parch Cabin Embarked
0         3    male  22.0      1      0   NaN        S
1         1  female  38.0      1      0   C85        C
2         3  female  26.0      0      0   NaN        S
3         1  female  35.0      1      0  C123        S
4         3    male  35.0      0      0   NaN        S
..      ...     ...   ...    ...    ...   ...      ...
886       2    male  27.0      0      0   NaN        S
887       1  female  19.0      0      0   B42        S
888       3  female   NaN      1      2   NaN        S
889       1    male  26.0      0      0  C148        C
890       3    male  32.0      0      0   NaN        Q

[891 rows x 7 columns]

In [10]:
y

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

In [11]:
x.isna().sum()

Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Cabin       687
Embarked      2
dtype: int64

In [12]:
#filing missing values

# for categorical
categorical_features = ["Cabin", "Embarked"]

si_categorical = SimpleImputer(strategy = "constant", fill_value = "Missing")

# for numerical

numerical_features = ["Age"]

si_numerical = SimpleImputer(strategy="mean")

#transforming them

transformer = ColumnTransformer(transformers= [("categorical_transformation", si_categorical, categorical_features),
                                               ("numerical_transformation", si_numerical, numerical_features)], remainder = "passthrough"
                               )

x_transformed = transformer.fit_transform(x)

In [13]:
x_transformed = pd.DataFrame(x_transformed, columns = ["Cabin", "Embarked", "Age","Pclass","Sex","Sibsp", "Parch"])

In [14]:
x_transformed.isna().sum()

Cabin       0
Embarked    0
Age         0
Pclass      0
Sex         0
Sibsp       0
Parch       0
dtype: int64

In [15]:
x_transformed["Cabin"].nunique()

148

In [16]:
for i in range(len(x_transformed["Cabin"])):
    x_transformed["Cabin"][i]=x_transformed["Cabin"][i][:1]

In [17]:



x_transformed.head()

Cabin Embarked Age Pclass     Sex Sibsp Parch
0     M        S  22      3    male     1     0
1     C        C  38      1  female     1     0
2     M        S  26      3  female     0     0
3     C        S  35      1  female     1     0
4     M        S  35      3    male     0     0

In [18]:
x_transformed["Cabin"].nunique()

9

In [19]:
x_transformed.dtypes

Cabin       object
Embarked    object
Age         object
Pclass      object
Sex         object
Sibsp       object
Parch       object
dtype: object

In [20]:
features=["Cabin", "Embarked","Sex"]

encoder = OneHotEncoder(handle_unknown = "ignore")

transformer2 = ColumnTransformer(transformers=[("OneHotEncoding", encoder, features)],
                                remainder="passthrough"
                                )
x_final = transformer2.fit_transform(x_transformed) 

In [21]:
x_final = pd.DataFrame(x_final)

In [22]:
x_final

0  1  2  3  4  5  6  7  8  9 10 11 12 13 14       15 16 17 18
0    0  0  0  0  0  0  0  1  0  0  0  0  1  0  1       22  3  1  0
1    0  0  1  0  0  0  0  0  0  1  0  0  0  1  0       38  1  1  0
2    0  0  0  0  0  0  0  1  0  0  0  0  1  1  0       26  3  0  0
3    0  0  1  0  0  0  0  0  0  0  0  0  1  1  0       35  1  1  0
4    0  0  0  0  0  0  0  1  0  0  0  0  1  0  1       35  3  0  0
..  .. .. .. .. .. .. .. .. .. .. .. .. .. .. ..      ... .. .. ..
886  0  0  0  0  0  0  0  1  0  0  0  0  1  0  1       27  2  0  0
887  0  1  0  0  0  0  0  0  0  0  0  0  1  1  0       19  1  0  0
888  0  0  0  0  0  0  0  1  0  0  0  0  1  1  0  29.6991  3  1  2
889  0  0  1  0  0  0  0  0  0  1  0  0  0  0  1       26  1  0  0
890  0  0  0  0  0  0  0  1  0  0  0  1  0  0  1       32  3  0  0

[891 rows x 19 columns]

In [23]:
classifier = LinearRegression()

In [24]:
classifier.fit(x_final, y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [25]:
t = np.array(x_final)

In [26]:
classifier.fit(t, y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [27]:
x_train,x_test,y_train,y_test =train_test_split(x_final, y , test_size= 0.6)

In [28]:
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((356, 19), (535, 19), (356,), (535,))

In [29]:
classifier2 = LinearRegression()

In [30]:
classifier2.fit(np.array(x_train), y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [31]:
y_pred = classifier2.predict(np.array(x_test)) 

In [32]:
from sklearn.ensemble import RandomForestClassifier

In [33]:
clf = RandomForestClassifier()

In [34]:
clf.fit(np.array(x_train), y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [35]:
y_pred = clf.predict((x_test))

In [36]:
y_pred

array([1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1,
       0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0,
       0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0,
       1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1,
       0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1,

In [37]:
from sklearn.metrics import accuracy_score

In [38]:

accuracy_score(y_test, y_pred)

0.7962616822429907

In [40]:
y_pred = pd.DataFrame(y_pred)

In [41]:
y_pred

0
0    1
1    0
2    0
3    0
4    1
..  ..
530  0
531  1
532  1
533  0
534  0

[535 rows x 1 columns]

In [42]:
y_pred.to_csv("Predicted_Titanic.csv", index = False)